atlante/atlante.go

```diff
@@ -21,31 +23,62 @@ import (
 )
 
 type ImgStruct struct {
-	Filename string
-	Width    int
-	Height   int
+	filename string
+	lck      sync.Mutex
+	// Have we generated the file?
+	// This will allow us to generate the file only when requested
+	generated bool
+	image     *image.Image
+	filestore filestore.FileWriter
+	Width     int
+	Height    int
 }
 
-func (img ImgStruct) ImageTagBase64() string {
-
-	return fmt.Sprintf(
-		`<image width="%d" height="%d" xlink:href="data:%s;base64,%s" />`,
-		img.Width,
-		img.Height,
-		img.MimeType(),
-		img.Base64Image(),
-	)
+// Filename returns the file name of the image
+func (img ImgStruct) Filename() (string, error) {
+	if img.generated {
+		return img.filename, nil
+	}
+	// We need to generate the file and then return the filename
+	return img.generateImage()
 }
 
-func (img ImgStruct) MimeType() string { return "image/png" }
-
-func (img ImgStruct) Base64Image() string {
-	filebytes, err := ioutil.ReadFile(img.Filename)
+// generateIamge is use to create the image into the filestore
+func (img *ImgStruct) generateImage() (fn string, err error) {
+	// Make sure generateImage is only called one at a time.
+	defer func() {
+		if err == nil {
+			img.generated = true
+		}
+	}()
+	if img.generated {
+		return img.filename, nil
+	}
+	if img.filestore == nil {
+		return img.filename, nil
+	}
+	file, err := img.filestore.Writer(img.filename, true)
 	if err != nil {
-		log.Printf("got error reading file %v: %v\n", img.Filename, err)
-		return ""
+		if err == filestore.ErrSkipWrite {
+			err = nil
```

Comment:
At first this didn't make sense to me but then I realized you're using this with the defer func. I think it's a bit complicated to follow. It should at least have a comment.

Reply:
Yeah, need to remove the `err=nil`, it's not needed and this is left over from when I was attempting to use a goroutine. I'm going clean this up. Looking at this morning it's too complicated.